src/StripeTests/Functional/TelemetryTest.cs

```diff
@@ -0,0 +1,216 @@
+                            m.Headers,
+                            (s) => s == "req_2",
+                            (d) => d >= 20)),
+                    ItExpr.IsAny<CancellationToken>());
+        }
+
+        [Fact]
+        public async Task TelemetryWorksWithConcurrentRequests()
+        {
+            this.ResetStripeClient();
+
+            var invocationCount = 0;
+            this.MockHttpClientFixture.MockHandler.Protected()
+                .Setup<Task<HttpResponseMessage>>(
+                    "SendAsync",
+                    ItExpr.IsAny<HttpRequestMessage>(),
+                    ItExpr.IsAny<CancellationToken>())
+                .Returns(async () =>
+                    {
+                        invocationCount += 1;
+
+                        await Task.Delay(10);
+
+                        var response = new HttpResponseMessage(HttpStatusCode.OK);
+                        response.Headers.Add("Request-Id", $"req_{invocationCount}");
+                        response.Content = new StringContent("{}");
+                        return response;
+                    });
+
+            StripeConfiguration.EnableTelemetry = true;
+            var service = new BalanceService();
+
+            // the first 2 requests will not contain telemetry
+            await Task.WhenAll(service.GetAsync(), service.GetAsync());
+
+            // the following 2 requests will contain telemetry
+            await Task.WhenAll(service.GetAsync(), service.GetAsync());
+
+            this.MockHttpClientFixture.MockHandler.Protected()
+                .Verify(
+                    "SendAsync",
+                    Times.Exactly(2),
+                    ItExpr.Is<HttpRequestMessage>(m =>
+                        !m.Headers.Contains("X-Stripe-Client-Telemetry")),
+                    ItExpr.IsAny<CancellationToken>());
+
+            this.MockHttpClientFixture.MockHandler.Protected()
+                .Verify(
+                    "SendAsync",
+                    Times.Exactly(2),
+                    ItExpr.Is<HttpRequestMessage>(m =>
+                        TelemetryHeaderMatcher(
+                            m.Headers,
+                            (s) => s == "req_1" || s == "req_2",
+                            (d) => d >= 10)),
+                    ItExpr.IsAny<CancellationToken>());
+        }
+
+        private static bool TelemetryHeaderMatcher(
+            HttpHeaders headers,
+            Func<string, bool> requestIdMatcher,
+            Func<long, bool> durationMatcher)
+        {
+            if (!headers.Contains("X-Stripe-Client-Telemetry"))
+            {
+                return false;
+            }
+
+            var payload = headers.GetValues("X-Stripe-Client-Telemetry").First();
+
+            var deserialized = JToken.Parse(payload);
+            var requestId = (string)deserialized["last_request_metrics"]["request_id"];
+            var duration = (long)deserialized["last_request_metrics"]["request_duration_ms"];
```

Comment:
I like that you're testing against the raw json instead of the `ClientTelemetryPayload`, as this ensures that we send the exact json that Stripe expects.

Reply:
I did it this way because `ClientTelemetryPayload` is `private` and I didn't want to expose it just for tests, but you're right that it actually makes the test better! :)